### PR TITLE
CUDA Top K / Top P Fixes

### PR DESCRIPTION
**Pull request by aciddelgado**
Top P and Top K on CUDA were not outputting expected results due to incorrect math. Specifically, we were:

1. Softmaxing vocab_size logits _before_ getting the top k subset when it should've been after.
2. Performing top p first, then top k. We should instead perform top p on the top k subset.

### CODE REVIEW

File: src/search.cpp [hunk 1]
```diff
@@ -217,11 +216,7 @@ void GreedySearch_Cpu::SampleTopKTopP(int k, float p, float temperature) {
     for (int i = 0; i < k; i++) {
       scores_top_k[i] = scores[indices[i]];
     }
-    if (k > 1) {
-      SoftmaxWithMax(scores_top_k, temperature, scores_top_k[0]);
-    } else {
-      scores_top_k[0] = 1.0f;
-    }
+    SoftmaxWithMax(scores_top_k, temperature, scores_top_k[0]);
```
Comment (RyanUnderhill) [inline src/search.cpp:219]: If k happens to be 1, does this still work fine?

**Review by RyanUnderhill — COMMENTED** [review]
What tests did you try to check the changes?

**Comment by aciddelgado** [comment]
> What tests did you try to check the changes?

Updated the tests to be more rigorous and have correct expected distributions

**Review by RyanUnderhill — APPROVED** [review]
